client/ayon_core/hosts/zbrush/api/lib.py

```diff
@@ -77,6 +77,20 @@ def get_current_file():
     return host.get_current_workfile()
 
 
+def execute_publish_model_with_dialog(filepath):
+    import time
+    save_file_zscript = ("""
+[IFreeze,
+[VarSet, filepath, "{filepath}"]
+[FileNameSetNext, #filepath]
+[IKeyPress, 13, [IPress, Tool:Export]]]
+[Sleep, 2]
+]
+""").format(filepath=filepath)
+    execute_zscript(save_file_zscript)
+    time.sleep(8)
```

Comment:
Just to elaborate on this with the info I also shared on discord, basically this:
```python
import time
start = time.time()
print(f"Start {time.time() - start}")
communicator.execute_zscript("[Sleep, 2]")
print(f"After first sleep {time.time() - start}")
communicator.execute_zscript("[Sleep, 2]")
print(f"After second sleep {time.time() - start}")
communicator.execute_zscript("[Sleep, 5]")
communicator.execute_zscript("[Sleep, 5]")
print(f"After two longer sleeps {time.time() - start}")
```

Should result in:
```
0.0
2.0
4.0
14.0
```
But then slightly higher numbers because of the time taken to execute, etc.
But it may not be lower numbers.

Since we only want to continue our python scripts as soon as the zscript finished completely with its execution.

Reply:
I feel like someone else at Ynput - who's done work on the After Effects / Photoshop integrations and likely has done some work on these communication servers with other hosts could bring some valuable insights to this as well from that experience.

Reply:
> Just to elaborate on this with the info I also shared on discord, basically this:
> 
> ```python
> import time
> start = time.time()
> print(f"Start {time.time() - start}")
> communicator.execute_zscript("[Sleep, 2]")
> print(f"After first sleep {time.time() - start}")
> communicator.execute_zscript("[Sleep, 2]")
> print(f"After second sleep {time.time() - start}")
> communicator.execute_zscript("[Sleep, 5]")
> communicator.execute_zscript("[Sleep, 5]")
> print(f"After two longer sleeps {time.time() - start}")
> ```
> 
> Should result in:
> 
> ```
> 0.0
> 2.0
> 4.0
> 14.0
> ```
> 
> But then slightly higher numbers because of the time taken to execute, etc. But it may not be lower numbers.
> 
> Since we only want to continue our python scripts as soon as the zscript finished completely with its execution.

Maybe one of the way we can do this is to introduce a while loop to check on the file existence. If you talk about exexute zscript is an issue. this means the issue is everywhere, including get current workfile, save workfile, open workfile, list instance, list container.

Reply:
> I feel like someone else at Ynput - who's done work on the After Effects / Photoshop integrations and likely has done some work on these communication servers with other hosts could bring some valuable insights to this as well from that experience.

The better reference is tvpaint, i have already mentioned that I would write dll plugins in additional info

Reply:
Here's an example 'workaround' as I described:
```python
import os
import tempfile

def execute_zscript_and_wait(zscript, wait=0.1, timeout=20):
    temp_path = os.path.join(
        tempfile.gettempdir(),
        "{tempfile.gettempprefix()}_{uuid}.txt"
    )
    temp_path = path.replace("\\", "/")
    assert not os.path.exists(temp_path)
    
    # At the end of the zscript write
    # temporary file to confirm the zscript
    # has run to the end of the script
    zscript += (f"""
[MemWriteString, EditMode, "1", 0]
[MemSaveToFile, EditMode, "{temp_path}", 0]
[MemDelete, EditMode]
]
""")
    execute_zscript(zscript)
    
    # Wait around until the zscript finished
    time_taken = 0
    while not os.path.exists(temp_path):
        time.sleep(wait)
        time_taken += wait
        if time_taken > timeout:
            raise RuntimeError(
                f"Timeout. Zscript took longer than "
                "{timeout}s to run."
            )
```

Probably not best performance wise - there may be other ways to get responses from Zbrush than this, but I think this should work at least.